packages/@uppy/core/src/Uppy.js

```diff
@@ -59,7 +59,7 @@ class Uppy {
       debug: false,
       restrictions: defaultRestrictionOptions,
       meta: {},
-      onBeforeFileAdded: (currentFile) => currentFile,
+      onBeforeFileAdded: (file, files) => !(file.id in files),
```

Comment:
```suggestion
      onBeforeFileAdded: (file, files) => !Object.keys(files).includes(file.id),
```

I think `in` operator can be a bit scary because it can give false positives for inherited properties like `__proto__`. for example:

```
'__proto__' in {}
```
returns true

Reply:
I don't think we need it, we control the `id` so it's always predictable. Meanwhile your version is significantly slower as it iterates over the entire array twice, which may be 10 or 20K files. This version is a quick look up table.

Reply:
Not sure if we always control ID, for example here:
https://github.com/transloadit/uppy/blob/fa19db7858069a57046c5e731921d07c01e46ded/packages/%40uppy/core/src/Uppy.js#L192

maybe we should add a warning in our API docs that the developer should be careful to sanitise the IDs that they provide to uppy's API.

maybe in the future we can rewrite to use `Map` instead, as it's fast and safe

Reply:
You can technically override it but we document no case for that. It's also high unlikely anyone will put an inherited property name as an `id`. I think we're fine here.

Reply:
it's unlikely that a normal user or developer will put the inherited property name there, but an *attacker* who is looking for an exploit is highly likely to put an inherited property there, so I like to code defensively and either
- make sure that nobody is able to do that
- warn developers that they need to guard against it in their end-user application

Reply:
What about using `Object.hasOwn`?